src/Specular/Internal/Incremental.purs

```diff
@@ -0,0 +1,410 @@
+  value <- runEffectFn1 source.compute node
+  runEffectFn2 Node.set_value node value
+
+disconnect :: forall a. EffectFn1 (Node a) Unit
+disconnect = mkEffectFn1 \node -> do
+  source <- runEffectFn1 Node.get_source node
+
+  dependencies <- source.dependencies
+  runEffectFn2 Array.iterate dependencies $ mkEffectFn1 \dependency -> do
+    runEffectFn2 removeDependent dependency (toSomeNode node)
+
+-- * Recompute
+
+stabilize :: Effect Unit
+stabilize = do
+--  trace "stabilize begin"
+  oldStabilizationNum <- runEffectFn1 Ref.read globalLastStabilizationNum 
+  let currentStabilizationNum = oldStabilizationNum + 1
+  runEffectFn2 Ref.write globalLastStabilizationNum currentStabilizationNum
+  runEffectFn2 Ref.write globalCurrentStabilizationNum currentStabilizationNum 
+
+  runEffectFn2 PQ.drain globalRecomputeQueue $ mkEffectFn1 \node -> do
+
+    height <- runEffectFn1 Node.get_height node
+    adjustedHeight <- runEffectFn1 Node.get_adjustedHeight node
+
+    if adjustedHeight > height then do
+--      trace $ "stabilize: node " <> show name <> ": height bump " <> show height <> " -> " <> show adjustedHeight
+
+      dependents <- runEffectFn1 Node.get_dependents node
+      runEffectFn2 MutableArray.iterate dependents $ mkEffectFn1 \dependent -> do
+        runEffectFn2 ensureHeight dependent (adjustedHeight + 1)
+
+      runEffectFn2 Node.set_height node adjustedHeight
+
+      -- Reconsider the node with new height
+      _ <- runEffectFn2 PQ.add globalRecomputeQueue node
+      pure unit
+
+    else do
+--      trace $ "stabilize: node " <> show name <> ": compute at height " <> show height
+
+      source <- runEffectFn1 Node.get_source node
+      -- oldValue_opt <- runEffectFn1 Node.get_value node
+      newValue_opt <- runEffectFn1 source.compute node
+
+      if Optional.isSome newValue_opt
+        -- && shouldNotCutOff oldValue_opt newValue
+      then do
+        let newValue = Optional.fromSome newValue_opt
+        runEffectFn2 Node.set_value node (Optional.some newValue)
+        runEffectFn2 Node.set_changedAt node currentStabilizationNum
+        
+        dependents <- runEffectFn1 Node.get_dependents node
+        runEffectFn2 MutableArray.iterate dependents $ mkEffectFn1 \dependent -> do
+          added <- runEffectFn2 PQ.add globalRecomputeQueue dependent
+--          if added then do
+--            dependentName <- runEffectFn1 Node.name dependent
+--            trace $ "stabilize: node " <> show dependentName <> " added to recompute queue"
+--          else do
+--            dependentName <- runEffectFn1 Node.name dependent
+--            trace $ "stabilize: node " <> show dependentName <> " already in recompute queue"
+          pure unit
+
+        observers <- runEffectFn1 Node.get_observers node
+        runEffectFn2 MutableArray.iterate observers $ mkEffectFn1 \observer -> do
+          -- FIXME: should be done outside stabilize loop, to avoid interfering with the process
```

Comment:
What about this `FIXME` ? Have this been resolved ?

Reply:
Yes, in Specular proper there's an additional queue for effects. Will remove comment.